framework/src/loops/ComputeMortarFunctor.C

```diff
@@ -82,175 +80,227 @@ ComputeMortarFunctor::operator()()
-        reinit_primary_elem, primary_side_id, _primary_boundary_id, TOLERANCE, &custom_xi2_pts);
-
-    // reinit neighbor materials, but be careful not to execute stateful materials since
-    // conceptually they don't make sense with mortar (they're not interpolary)
-    _fe_problem.reinitMaterialsNeighbor(primary_ip->subdomain_id(),
-                                        /*tid=*/0,
-                                        /*swap_stateful=*/false,
-                                        /*execute_stateful=*/false);
-
-    // reinit the variables/residuals/jacobians on the lower dimensional element corresponding to
-    // the secondary face. This must be done last after the dof indices have been prepared for the
-    // secondary (element) and primary (neighbor)
-    _subproblem.reinitLowerDElem(secondary_face_elem, /*tid=*/0, &custom_xi1_pts);
-
-    // All this does currently is sets the neighbor/primary lower dimensional elem in Assembly and
-    // computes its volume for potential use in the MortarConstraints. Solution continuity
-    // stabilization for example relies on being able to access the volume
-    _subproblem.reinitNeighborLowerDElem(msinfo.primary_elem);
-
-    // reinit higher-dimensional secondary face/boundary materials. Do this after we reinit lower-d
-    // variables in case we want to pull the lower-d variable values into the secondary
-    // face/boundary materials. Be careful not to execute stateful materials since conceptually they
-    // don't make sense with mortar (they're not interpolary)
-    _fe_problem.reinitMaterialsFace(secondary_ip->subdomain_id(),
-                                    /*tid=*/0,
-                                    /*swap_stateful=*/false,
-                                    /*execute_stateful=*/false);
-    _fe_problem.reinitMaterialsBoundary(
-        _secondary_boundary_id, /*tid=*/0, /*swap_stateful=*/false, /*execute_stateful=*/false);
-
-    num_cached++;
-    if (!_fe_problem.currentlyComputingJacobian())
-    {
-      for (auto * const mc : _mortar_constraints)
-      {
-        mc->setNormals(mc->interpolateNormals() ? normals : nodal_normals);
-        mc->computeResidual();
-      }
+        for (auto * const mc : _mortar_constraints)
+        {
+          mc->setNormals(mc->interpolateNormals() ? normals : nodal_normals);
+          mc->computeResidual();
+        }
 
-      _assembly.cacheResidual();
-      _assembly.cacheResidualNeighbor();
-      _assembly.cacheResidualLower();
+        _assembly.cacheResidual();
+        _assembly.cacheResidualNeighbor();
+        _assembly.cacheResidualLower();
 
-      if (num_cached % 20 == 0)
-        _assembly.addCachedResiduals();
-    }
-    else
-    {
-      for (auto * const mc : _mortar_constraints)
-      {
-        mc->setNormals(mc->interpolateNormals() ? normals : nodal_normals);
-        mc->computeJacobian();
+        if (num_cached % 20 == 0)
+          _assembly.addCachedResiduals();
```

Comment:
@lindsayad I've been puzzled about `num_cached % 20 == 0`: is this a heuristic optimization?

Reply:
I think the idea is that adding entries into the actual `NumericVector` holding the residual is expensive so we do it in batches. But whether this is truly an important expense is unknown to me.